### PR TITLE
check for stack size before executing the opcode

### PR DESCRIPTION
**Pull request by mubarak23**
- [x] issue #176 
- [x] follows contribution [guide](https://github.com/keep-starknet-strange/shinigami/blob/main/CONTRIBUTING.md)

### CODE REVIEW

File: src/engine.cairo [hunk 1]
```diff
@@ -50,9 +50,18 @@ pub trait EngineTrait {
...
     // Return the script since last OP_CODESEPARATOR
     fn sub_script(ref self: Engine) -> ByteArray;
+
+    // check the stack size before pushing data
+    fn check_stack_size(ref self: Engine) -> Result<bool, felt252>;
 }
 
 pub impl EngineImpl of EngineTrait {
+    fn check_stack_size(ref self: Engine) -> Result<bool, felt252> {
+        if self.dstack.len() >= 1000 {
```
Comment (b-j-roberts) [inline src/engine.cairo:60]: Can you make the 1000 a constant defined at the top of the file maybe, `MAX_STACK_SIZE`.

**Review by b-j-roberts — CHANGES_REQUESTED** [review]

**Comment by mubarak23** [comment]
@b-j-roberts changes request base on the review has been made.

File: src/engine.cairo [hunk 2]
```diff
@@ -110,11 +120,16 @@ pub impl EngineImpl of EngineTrait {
...
         if !self.cond_stack.branch_executing() && !flow::is_branching_opcode(opcode) {
+            // check stack size
+            self.check_stack_size()?;
             Opcode::is_opcode_disabled(opcode, ref self)?;
             self.opcode_idx += 1;
             return Result::Ok(true);
         }
 
+        // check stack size
+        self.check_stack_size()?;
```
Comment (b-j-roberts) [inline src/engine.cairo:131]: It might be better to check the stack size after executing, so moving it to like 135

File: src/engine.cairo [hunk 3]
```diff
@@ -110,11 +120,16 @@ pub impl EngineImpl of EngineTrait {
         Opcode::is_opcode_always_illegal(opcode, ref self)?;
 
         if !self.cond_stack.branch_executing() && !flow::is_branching_opcode(opcode) {
+            // check stack size
+            self.check_stack_size()?;
```
Comment (b-j-roberts) [inline src/engine.cairo:124]: We wouldn't need to check here if we move the `check_stack_size` to after execute.

File: src/engine.cairo [hunk 4]
```diff
@@ -50,9 +50,19 @@ pub trait EngineTrait {
     fn has_flag(ref self: Engine, flag: ScriptFlags) -> bool;
     // Return the script since last OP_CODESEPARATOR
     fn sub_script(ref self: Engine) -> ByteArray;
+
+    // check the stack size before pushing data
+    fn check_stack_size(ref self: Engine) -> Result<(), felt252>;
 }
+pub const MAX_STACK_SIZE: u32 = 1000;
```
Comment (b-j-roberts) [inline src/engine.cairo:57]: Maybe moving this to `src/stack.cairo` would be better.

File: src/engine.cairo [hunk 5]
```diff
@@ -50,9 +50,19 @@ pub trait EngineTrait {
...
 
 pub impl EngineImpl of EngineTrait {
+    fn check_stack_size(ref self: Engine) -> Result<(), felt252> {
+        if self.dstack.len() >= MAX_STACK_SIZE {
```
Comment (b-j-roberts) [inline src/engine.cairo:61]: It seems the max stack size check should include the alt stack : https://github.com/btcsuite/btcd/blob/2b53ed198955ac40fe5b3ce4a854e9f5bfa68258/txscript/engine.go#L971

**Review by b-j-roberts — CHANGES_REQUESTED** [review]

File: src/engine.cairo (hunk 5 above)
Comment (mubarak23) [reply src/engine.cairo:61]: so both astack and dstack combine, must be less than 1000 ?

File: src/engine.cairo (hunk 2 above)
Comment (mubarak23) [reply src/engine.cairo:131]: any reason for this, i though, it will be better we check the stack size before executing the opcode

**Review by b-j-roberts — APPROVED** [review]
Thank you!

**Pull request merged**